lib/thread.c

```diff
@@ -1669,10 +1732,48 @@ void thread_call(struct thread *thread)
 	unsigned long helper;
 #endif
 	RUSAGE_T before, after;
+	struct timeval wkptime, diff;
+	uint64_t wkpintvl_usecs;
 
 	GETRUSAGE(&before);
 	thread->real = before.real;
 
+	if (thread_collect_stats) {
+		/*
+		 * The following code measures interval between consecutive
+		 * wakeup of the various threads under the current process
+		 * and log that in order to detect any possible CPU starvations.
+		 * Offcourse it can not be used to differentiate between
+		 * wether a process was indeed ready but it could not be
+		 * scheduled (due to other process hogging CPU) and that of
+		 * the process really being idle for long.
+		 *
+		 * In cases if a deamon is supposed to run some timely regular
+		 * jobs, the data captured here can definitely highlight if
+		 * the process ever went through CPU starvation or not.
+		 */
+		(void)monotime_to_realtime(&before.real, &wkptime);
+		frr_with_mutex (&thread->master->mtx) {
```

Comment:
again, I don't quite get why the mutex has to be held here - we're running in the context of pthread whose data this is?

Reply:
@mjstapp Please note that the 'thread' being passed to thread_call() is actually a copy of the actual 'thread' being maintained under the masters. But I also realize that all the threads under a thread_master are running under the same pthread context. So I think we can avoid this mutex lock here. If someone can confirm we don't need the lock here I can happily remove it. But we will definitely need the lock in thread_dump_master_statisitics() since that is walking over all thread masters from a single pthread context.